src/realm/sync/noinst/pending_reset_store.hpp

```diff
@@ -0,0 +1,91 @@
+///////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2024 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_NOINST_PENDING_RESET_STORE_HPP
+#define REALM_NOINST_PENDING_RESET_STORE_HPP
+
+#include <realm/status.hpp>
+#include <realm/timestamp.hpp>
+#include <realm/sync/client_base.hpp>
+#include <realm/sync/config.hpp>
+#include <realm/sync/protocol.hpp>
+
+#include <realm/sync/noinst/sync_metadata_schema.hpp>
+
+#include <ostream>
+#include <optional>
+
+namespace realm::sync {
+
+struct PendingReset {
+    using Action = sync::ProtocolErrorInfo::Action;
+    Timestamp time;
+    ClientResyncMode mode;
+    Action action = Action::NoAction;
+    std::optional<Status> error;
+};
+
+std::ostream& operator<<(std::ostream& os, const sync::PendingReset& pr);
+bool operator==(const sync::PendingReset& lhs, const sync::PendingReset& rhs);
+bool operator==(const sync::PendingReset& lhs, const PendingReset::Action& action);
+
+class PendingResetStore {
```

Comment:
I find it a bit odd that all methods are static in this class, and you create temporary instances used in some of these methods.

Reply:
The class instances are just for holding the table/column key variables after the schema metadata info is loaded. I had gone back and forth with this and think the static functions provides the caller control over the transactions and when the changes are committed (e.g. remove old and add new pending reset info in a single transaction, like was being done before). Although it did require some updates to the SchemaVersion classes and how they handle transactions.

Plus, this class is only used in a very small window and it seemed excessive to keep an object (and dbref) around during every SessionWrapper lifetime.

Reply:
right. I wasn't suggesting using a store object, but maybe the same old static functions if possible.

Reply:
I guess I could pull the static functions outside the class and retain the old namespace for them... Not sure if it's necessary, since this way everything is contained in a single unit/class and it incorporates the internal object for holding the schema info.

Reply:
fwiw, I don't think its worth refactoring this again just to use the old function names.